### PR TITLE
db rollback in case of panic

### LINKED ISSUE #1454: Panic database transaction should be rollbacked
Database transaction uses a callback to perform a scope operation during the database transaction.
In case of this function panic, we currently do not rollback the transaction which can cause resources such as locked to be kept (at least until database timeout).

### PR DESCRIPTION
**Pull request by nopcoder**
Fix #1454

### CODE REVIEW

**Comment by ozkatz** [comment]
@nopcoder if a panic happens while a tx is open, it will not get committed, so why explicitly roll it back?

**Comment by nopcoder** [comment]
> @nopcoder if a panic happens while a tx is open, it will not get committed, so why explicitly roll it back?

We don't need to have panic in the code, but we also need to abort open transactions.
In the case of the above, a test caused the panic and the out metadata locked transaction was left open, the next test that asked for out application-level locking was waiting until the test timed out.

**Comment by nopcoder** [comment]
> @nopcoder if a panic happens while a tx is open, it will not get committed, so why explicitly roll it back?

I'll add a test code

File: db/database.go [hunk 1]
```diff
@@ -162,6 +162,13 @@ func (d *PgxDatabase) Transact(fn TxFunc, opts ...TxOpt) (interface{}, error) {
 	}
 	var attempt int
 	var ret interface{}
+	var tx pgx.Tx
+	defer func() {
+		if p := recover(); p != nil && tx != nil {
+			_ = tx.Rollback(options.ctx)
```
Comment (arielshaqed) [inline db/database.go:168]: Consider reporting this error, e.g.
```suggestion
			err = tx.Rollback(options.ctx)
			if err != nil {
				p = fmt.Errorf("%w, and %s while rolling back", p, err)
			}
```
If something *does* go wrong rolling back during a panic, more information is better.

File: db/database.go [hunk 2]
```diff
@@ -162,6 +162,13 @@ func (d *PgxDatabase) Transact(fn TxFunc, opts ...TxOpt) (interface{}, error) {
 	}
 	var attempt int
 	var ret interface{}
+	var tx pgx.Tx
+	defer func() {
```
Comment (arielshaqed) [inline db/database.go:166]: Can you just move this to after we open the transaction, below?  That way we can just use tx, and I think the code becomes more clear.

File: db/database.go [hunk 3]
```diff
@@ -162,6 +162,13 @@ func (d *PgxDatabase) Transact(fn TxFunc, opts ...TxOpt) (interface{}, error) {
...
+		if p := recover(); p != nil && tx != nil {
+			_ = tx.Rollback(options.ctx)
+			panic(p)
```
Comment (arielshaqed) [inline db/database.go:169]: Does this lose the panic stacktrace (or is it stored on `p`?)?  If so, can we keep it somehow?

**Review by arielshaqed — APPROVED** [review]
Nice!
Makes sense, and should help our tests fail more cleanly.

File: db/database.go (hunk 2 above)
Comment (nopcoder) [reply db/database.go:166]: using defer in a loop can be a problem - we would like to have this one once or scope the operation into a different function and make sure it is committed there

File: db/database.go (hunk 3 above)
Comment (nopcoder) [reply db/database.go:169]: Verified that the stack trace captures the original panic stack

File: db/database.go (hunk 1 above)
Comment (nopcoder) [reply db/database.go:168]: Currently, we do no log rollback in the Transaction function. We can capture the information and the stack trace in the top-level handler if needed.
But what will be the added value of the rollback error as we assume everything is aborted in this flow.

**Pull request merged**